File: haskell/day22.hs
import qualified Data.Either as E
import qualified Text.ParserCombinators.Parsec as PCPS

eol = PCPS.char '\n'

inputHeader = do
    PCPS.string "Player " >> PCPS.digit >> PCPS.char ':' >> eol

inputCard = do
    card <- (PCPS.many1 PCPS.digit) <* eol
    return (read card :: Int)

inputCardList =
    PCPS.many1 inputCard

inputFile = do
    inputHeader
    player1 <- inputCardList
    eol
    inputHeader
    player2 <- inputCardList
    PCPS.eof
    return (player1, player2)

score :: [Int] -> Int
score hand =
    sum $ map (\(c, n) -> c * n) $ zip (reverse hand) [1..]

game :: [Int] -> [Int] -> [Int]
game hand1 hand2 =
    if null hand1 || null hand2 then
        if null hand1 then hand2 else hand1
    else
        game hand1' hand2'
        where
            card1:remaining1 = hand1
            card2:remaining2 = hand2
            winPlayer1 = card1 >= card2
            cards = if winPlayer1 then [card1, card2] else [card2, card1]
            hand1' = remaining1 ++ if winPlayer1 then cards else []
            hand2' = remaining2 ++ if not winPlayer1 then cards else []

answer :: ([Int], [Int]) -> Int
answer (hand1, hand2) = score $ game hand1 hand2

main :: IO ()
main = do
    let fn = "../input/day22.txt"
    inputText <- readFile fn
    let input = E.fromRight undefined $ PCPS.parse inputFile fn inputText
    putStrLn $ show $ answer $ input

File: c++/day22.cpp
#include <cstdlib>
#include <deque>
#include <fstream>
#include <iostream>
#include <set>
#include <string>

using Card = int;
using Hand = std::deque<Card>;
struct Hands {
    Hand player1;
    Hand player2;
};

Hands readParseInput(std::string fileName) {
    std::ifstream file(fileName);
    Hands hands;

    std::string line;
    // Player 1:
    std::getline(file, line);
    while (std::getline(file, line)) {
        if (!line.length())
            break;
        hands.player1.emplace_back(atoi(line.c_str()));
    }
    // Player 2:
    std::getline(file, line);
    while (std::getline(file, line))
        hands.player2.emplace_back(atoi(line.c_str()));
    file.close();

    return hands;
}

bool playGame(Hands &hands) {
    while (true) {
        // Player 1 wins
        if (!hands.player2.size())
            return true;
        // Player 2 wins
        if (!hands.player1.size())
            return false;
        Card card1 = hands.player1.front();
        hands.player1.pop_front();
        Card card2 = hands.player2.front();
        hands.player2.pop_front();
        if (card1 > card2) {
            hands.player1.push_back(card1);
            hands.player1.push_back(card2);
        } else {
            hands.player2.push_back(card2);
            hands.player2.push_back(card1);
        }
    }
}

int answer(Hands hands) {
    bool player1Won = playGame(hands);
    Hand &winningHand = player1Won ? hands.player1 : hands.player2;
    int len = winningHand.size();
    int answer = 0;
    for (const auto &card : winningHand)
        answer += len-- * card;
    return answer;
}

int main(void) {
    std::cout << answer(readParseInput("../input/day22.txt")) << '\n';
    return 0;
}
